### DISCUSSION TITLE
Large libraries: use common CDNs or purge unused code?

### DISCUSSION THREAD

**Discussion by artoonie** [GoogleChrome/lighthouse-ci#518] [Q&A] [answered]
It seems large libraries are penalized for long download times on slow networks. The recommendation is to remove unused CSS/JS.

Is it better to assume users have downloaded and cached common CDN-hosted files, or does lighthouse always show the worst-case scenario with no cache?

**Comment by patrickhulce** [comment, marked as answer]
> Is it better to assume users have downloaded and cached common CDN-hosted files

In practice this is not an assumption that is true. In fact, browsers by now are moving to a [double-keyed cache](https://github.com/whatwg/fetch/issues/904) where no CDN assets are shared between sites at all, even if the user had the unlikely exact combination of library versions from the same CDN as another recently visited site.

Also worth noting a significant portion of the cost isn't just downloading a large JS library, but in parsing and initial evaluation as well which does not receive the same benefits of the network disk cache.

> does lighthouse always show the worst-case scenario with no cache?

It shows the no cache performance by default. `--collect.settings.disableStorageReset` combined with a `--collect.puppeterScript` can measure the warm load too.

**Reply by artoonie** [reply]
Thank you for the educational comment! I had not heard of double-keyed cache, good to know. I thought the cost was limited to just the parsing of the JS libraries. 🙏
